src/game/gamemain.ts

```diff
@@ -0,0 +1,318 @@
+            });
+          }
+        } else {
+          this.debris.push({
+            pos: pos,
+            sprite: new Sprite('https://sapcey.netlify.app/img/sprites.png', [ 0, 129 ], [ 187, 85 ])
+          });
+        }
+      }
+
+      this.score++;
+
+      if (this.col > 100) {
+        this.isGameOver = true;
+      }
+
+      this.checkCollisions();
+      this.setCollisions(this.col);
+      this.setScore(this.score);
+      this.setGameOverStatus(this.isGameOver);
+      this.setGamePauseStatus(this.isGamePaused);
+    }
+  }
+
+  updateEntities(dt: number): void {
+    // Update the player sprite animation
+    if (this.player.side === 'right' && this.player.state === 'shift') {
+      if (this.player.pos[0] < this.canvas.width / 2 + 102 - 17) {
+        this.player.pos[0] += this.playerSpeed * dt;
+      } else {
+        this.player.state = 'flight';
+      }
+    } else if (this.player.side === 'left' && this.player.state === 'shift') {
+      if (this.player.pos[0] > this.canvas.width / 2 - 102 - 17) {
+        this.player.pos[0] -= this.playerSpeed * dt;
+      } else {
+        this.player.state = 'flight';
+      }
+    }
+    this.player.sprite.update(dt);
+
+    // Update all walls
+    for (let i = 0; i < this.walls.length; i++) {
+      this.walls[i].pos[1] += this.wallsSpeed * dt;
+      this.walls[i].sprite.update(dt);
+
+      // Remove if offscreen
+      if (this.walls[i].pos[1] > this.canvas.height) {
+
+        this.walls[i].pos[1] = 0 - 96;
+      }
+    }
+
+    // Update all debris
+    for (let i = 0; i < this.debris.length; i++) {
+      this.debris[i].pos[1] += this.debrisSpeed * dt;
+      this.debris[i].sprite.update(dt);
+
+      // Remove if offscreen
+      if (this.debris[i].pos[1] > this.canvas.height) {
+        this.debris.splice(i, 1);
+        i--;
+      }
+    }
+  }
+
+  collides(x: number, y: number, r: number, b: number, x2: number, y2: number, r2: number, b2: number): boolean {
+    return !(r <= x2 || x > r2
+        || b <= y2 || y > b2);
+  }
+
+  boxCollides(pos: [number, number], size: [number, number], pos2: [number, number], size2: [number, number]): boolean {
+    return this.collides(pos[0], pos[1],
+      pos[0] + size[0], pos[1] + size[1],
+      pos2[0], pos2[1],
+      pos2[0] + size2[0], pos2[1] + size2[1]);
+  }
+
+  checkPlayerBounds(): void {
+    // Check bounds
+    if (this.player.pos[0] < 0) {
+      this.player.pos[0] = 0;
+    } else if (this.player.pos[0] > this.canvas.width - this.player.sprite.size[0]) {
+      this.player.pos[0] = this.canvas.width - this.player.sprite.size[0];
+    }
+
+    if (this.player.pos[1] < 0) {
+      this.player.pos[1] = 0;
+    } else if (this.player.pos[1] > this.canvas.height - this.player.sprite.size[1]) {
+      this.player.pos[1] =this.canvas.height - this.player.sprite.size[1];
+    }
+  }
+
+  checkCollisions(): void {
+    this.checkPlayerBounds();
+
+    for (let i = 0; i < this.debris.length; i++) {
```

Comment:
❗️
Здесь и везде стоит использовать методы массивов, если нет очень веской причины сбиваться в императивщину
```suggestion
    this.debris.forEach(...)
```

Reply:
Тут скорее всего будет все меняться, так что пока не трогаю)